Sources/Datadog/RUM/RUMTelemetry.swift

```diff
@@ -105,53 +99,58 @@ internal final class RUMTelemetry: Telemetry {
-            let actionId = context.activeUserActionID?.toRUMDataFormat
-            let viewId = context.activeViewID?.toRUMDataFormat
-            let sessionId = context.sessionID == RUMUUID.nullUUID ? nil : context.sessionID.toRUMDataFormat
+            let attributes = context.featuresAttributes["rum"]
+
+            let applicationId = attributes?["application_id", type: String.self]
+            let sessionId = attributes?["session_id", type: String.self]
+            let viewId = attributes?["view.id", type: String.self]
+            let actionId = attributes?["user_action.id", type: String.self]
 
             let event = TelemetryErrorEvent(
                 dd: .init(),
                 action: actionId.map { .init(id: $0) },
-                application: .init(id: self.applicationID),
-                date: date.timeIntervalSince1970.toInt64Milliseconds,
+                application: applicationId.map { .init(id: $0) },
+                date: date.addingTimeInterval(context.serverTimeOffset).timeIntervalSince1970.toInt64Milliseconds,
                 experimentalFeatures: nil,
                 service: "dd-sdk-ios",
                 session: sessionId.map { .init(id: $0) },
-                source: TelemetryErrorEvent.Source(rawValue: self.source) ?? .ios,
+                source: .init(rawValue: context.source) ?? .ios,
                 telemetry: .init(error: .init(kind: kind, stack: stack), message: message),
-                version: self.sdkVersion,
+                version: context.sdkVersion,
                 view: viewId.map { .init(id: $0) }
             )
 
             writer.write(value: event)
         }
     }
 
-    private func record(event id: String, operation: @escaping (RUMContext, Writer) -> Void) {
-        let rum = core.v1.feature(RUMFeature.self)
-
+    private func record(event id: String, operation: @escaping (DatadogContext, Writer) -> Void) {
         guard
-            sampler.sample(),
-            let monitor = Global.rum as? RUMMonitor,
-            let writer = rum?.storage.writer
+            let rum = core.v1.scope(for: RUMFeature.self),
+            sampler.sample()
         else {
             return
         }
 
-        monitor.contextProvider.async { context in
+        rum.eventWriteContext { context, writer in
             // reset recorded events on session renewal
-            if context.sessionID != self.currentSessionID {
-                self.currentSessionID = context.sessionID
-                self.eventIDs = []
-            }
-
-            // record up de `MaxEventsPerSessions`, discard duplicates
-            if self.eventIDs.count < RUMTelemetry.MaxEventsPerSessions, !self.eventIDs.contains(id) {
-                self.eventIDs.insert(id)
-                operation(context, writer)
+            let attributes = context.featuresAttributes["rum"]
+            let sessionId = attributes?["session_id", type: String.self]
+
+            self.queue.sync {
```

Comment:
Do we need `.sync {}` here 🤔💭? I can't see why it can be `.async {}` in `RemoteLogger`, but `.sync {}` here. Hope it's a leftover 🙏🙂, no?

Reply:
Good catch!